traffic_router/core/src/test/java/org/apache/traffic_control/traffic_router/core/external/BufferedResponseTest.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.catalina.LifecycleException;
+import org.apache.http.Header;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpHead;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.util.EntityUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.io.IOException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.junit.Assert.assertThat;
+
+@Category(ExternalTest.class)
+public class BufferedResponseTest {
+	final private String routerHttpPort = System.getProperty("routerHttpPort", "8888");
+	private CloseableHttpClient httpClient;
+
+	@Before
+	public void before() throws LifecycleException {
+		httpClient = HttpClientBuilder.create().build();
+	}
+
+	@After
+	public void after() throws Exception {
+		if (httpClient != null) httpClient.close();
+	}
+
+	@Test
+	public void itSetsContentLengthHeaderFor404() throws IOException {
+		final String encodedUrl = URLEncoder.encode("http://trafficrouter01.somedeliveryservice.somecdn.domain.foo/stuff", "utf-8");
+		final HttpGet httpGet = new HttpGet("http://localhost:3333/crs/deliveryservices?url=" + encodedUrl);
+		CloseableHttpResponse response = null;
+
+		try {
+			response = httpClient.execute(httpGet);
+			assertThat(response.getStatusLine().getStatusCode(), equalTo(404));
+			assertThat(response.getFirstHeader("Transfer-Encoding"), nullValue());
+			assertThat(response.getFirstHeader("Content-Length"), notNullValue());
+		} finally {
+			if (response != null) response.close();
+		}
+	}
+
+	@Test
+	public void itSetsTheSameContentLengthForHeadAndGet() throws IOException {
+		final List<String> paths = new ArrayList<>();
+		paths.add("http://localhost:3333/crs/stats");
+		paths.add("http://localhost:3333/crs/locations/caches");
+		paths.add("http://localhost:3333/crs/consistenthash/deliveryservice?deliveryServiceId=csd-target-1&requestPath=/");
+
+		for (final String path : paths) {
+			final List<HttpRequestBase> requests = new ArrayList<>();
+			requests.add(new HttpHead(path));
+			requests.add(new HttpGet(path));
+
+			final List<Integer> contentLengths = new ArrayList<>();
+
+			for (final HttpRequestBase request : requests) {
+				CloseableHttpResponse response = null;
+
+				try {
+					response = httpClient.execute(request);
+					final Header contentLengthHeader = response.getFirstHeader("Content-Length");
+
+					assertThat(response.getFirstHeader("Transfer-Encoding"), nullValue());
```

Comment:
same as the above 2

Reply:
Using `org.springframework.http.HttpHeaders.CONTENT_LENGTH` in faafed127b